src/Workspaces/Core/Portable/Workspace/IsolatedAnalyzerReferenceSet.Core.cs

```diff
@@ -37,46 +39,46 @@ internal sealed partial class IsolatedAnalyzerReferenceSet
     /// </summary>
     private static readonly Dictionary<Checksum, WeakReference<IsolatedAnalyzerReferenceSet>> s_checksumToReferenceSet = [];
 
-    private static int s_sweepCount = 0;
-
     /// <summary>
-    /// Final set of <see cref="AnalyzerReference"/> instances that will be passed through the workspace down to the compiler.
+    /// The current isolated reference set we're trying to use to load analyzers in.  We'll keep using the same set
+    /// until we run into a conflict that prevents it from being used.  At that point we'll create a new set and use
+    /// that one from that point on (and so on).  The old sets will stay alive as long as any AnalyzerReference (or
+    /// ISourceGenerator or DiagnosticAnalyzer from it) is alive.  Once all of those are garbage collected, the set
+    /// itself can be collected.  At that point it will release it's assembly load context, freeing everything.
     /// </summary>
-    public ImmutableArray<AnalyzerReference> AnalyzerReferences { get; }
+    /// <remarks>
+    /// To determine if we have a conflict, we keep track of the mvid of each <see cref="AnalyzerFileReference"/> when
+    /// the set was created.  When trying to reuse the set, we see if any of the references we now have has a different
+    /// mvid from that creation point.  If so, we have a conflict and we make a new set.
+    /// </remarks>
+    private static IsolatedAnalyzerReferenceSet? s_lastCreatedAnalyzerReferenceSet;
```

Comment:
This is also guarded by the SemaphoreSlim gate too? If so, add a comment (or update the comment on the gate)

Reply:
doc'ed all oft his.